File: src/piece_map.rs
//! Piece-map column aggregation used by the Qt `PieceMap` widget.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PieceFill {
    Empty,
    Partial,
    Have,
}

pub fn piece_column_fill(have: &[bool], width: usize) -> Vec<PieceFill> {
    if width == 0 {
        return Vec::new();
    }
    let total = have.len();
    (0..width)
        .map(|column| {
            if total == 0 {
                return PieceFill::Empty;
            }
            let start = column * total / width;
            let end = (start + 1).max((column + 1) * total / width);
            let chunk = &have[start..end.min(total)];
            if chunk.is_empty() {
                PieceFill::Empty
            } else if chunk.iter().all(|&bit| bit) {
                PieceFill::Have
            } else if chunk.iter().any(|&bit| bit) {
                PieceFill::Partial
            } else {
                PieceFill::Empty
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_have_is_empty_columns() {
        assert_eq!(
            piece_column_fill(&[], 4),
            vec![
                PieceFill::Empty,
                PieceFill::Empty,
                PieceFill::Empty,
                PieceFill::Empty
            ]
        );
    }

    #[test]
    fn all_have_fills_every_column() {
        assert!(piece_column_fill(&[true, true, true, true], 8)
            .iter()
            .all(|fill| *fill == PieceFill::Have));
    }

    #[test]
    fn mixed_bits_produce_partial_columns() {
        let fills = piece_column_fill(&[true, false], 1);
        assert_eq!(fills, vec![PieceFill::Partial]);
    }
}

File: cpp/piece_map.cpp
#include "piece_map.hpp"

namespace i2p {

QVector<PieceFill> pieceColumnFill(const QVector<bool> &have, int width)
{
    if (width <= 0) {
        return {};
    }
    const int total = have.size();
    QVector<PieceFill> columns;
    columns.reserve(width);
    for (int column = 0; column < width; ++column) {
        if (total == 0) {
            columns.push_back(PieceFill::Empty);
            continue;
        }
        const int start = column * total / width;
        const int end = qMax(start + 1, (column + 1) * total / width);
        bool any = false;
        bool all = true;
        for (int index = start; index < end && index < total; ++index) {
            any = any || have[index];
            all = all && have[index];
        }
        if (!any) {
            columns.push_back(PieceFill::Empty);
        } else if (all) {
            columns.push_back(PieceFill::Have);
        } else {
            columns.push_back(PieceFill::Partial);
        }
    }
    return columns;
}

} // namespace i2p
